### PR TITLE
3847 - Add toggle icon and children count with tree

### LINKED ISSUE #3847: Tree: Separate button icon to expand/collapse node + node children count awareness & display
As shown in the screenshot below, we would like to have a separate icon which can be clicked to collapse/expand the node and the support for children count feature. In addition we would like to support an additional icon.

tree_node_separate_expand_collapse_button_icon

**Describe the solution you'd like**

**Expand/Collapse (+/-) icons:**
- A configuration to show/hide separate button icons for expand/collapse.
- The expand/collapse icon should not be shown for a leaf node
- Click on expand/collapse button icon should only expand/collapse the tree node.
- Click on expand/collapse button should NOT select the tree node.
- Good to have: The possibility to use different expand/collapse icons

**Count:**
- The children are fetched when user clicks on the node. We will know the children count but won't have the exact children available until user clicks on the node and the server returns the children. So, tree should have some mechanism to show expand/collapse icon on the basis of count. 
- Each tree node should have some property like count of children (auto calculated on the basis of actual children) or directly set via a tree node property. The count label is shown beside the name. This is similar to "N results" feature on datagrid.
- The label automatically changes if the number of children changes (by addition or removal) if the count is auto-calculated instead of manually being set via tree node property.

**Angular support:**
- The feature should also be available in enterprise-ng

**Additional context**
Infor PIM
Product Owner: @karinthulin

### PR DESCRIPTION
**Pull request by deep7102**
**Explain the _details_ for making this change. What existing problem does the pull request solve?**
Added the ability to have separate icon button for expand/collapse and children count with Tree.

**Related github/jira issue (required)**:
Closes #3847

**Steps necessary to review your pull request (required)**:
Pull this branch and build/run the demo app

http://localhost:4000/components/tree/test-toggle-icon-and-count.html
- Should see plus-minus icon to parent nodes
- Should only do expand/collapse by icon (not by clicking on node text)
- Should change plus-minus icon based on parent node expand/collapse
- If not given custom icon then should use default folder icon (see node with name `Road`)
- Should show children count beside the node name text

http://localhost:4000/components/tree/test-toggle-icon-and-count.html
- Open node `Mountain (6)` by clicking icon
- Select node `Parts (39)` by clicking on text
- Open then close (expand/collapse) node `Foldable (1)` by clicking icon
- Now Close node `Mountain (6)` by clicking icon
- See node `Parts (39)` should remain selected

http://localhost:4000/components/tree/test-toggle-icon-and-count-ajax.html
- Should see plus-minus icon to parent node
- Should only do expand/collapse by icon (not by clicking on node text)
- Should show children count beside the node name text
- Should load async children soon click to parent node `Road` 

http://localhost:4000/components/tree/test-toggle-icon-and-count-sortable.html
- Should see plus-minus icon to parent nodes 
- Should only do expand/collapse by icon (not by clicking on node text)
- Should show children count beside the node name text
- Should update children count soon rearrange the children nodes by drag/drop

http://localhost:4000/components/tree/test-toggle-icon-and-count-markup.html
http://localhost:4000/components/tree/test-toggle-icon-and-count-badges.html
http://localhost:4000/components/tree/test-toggle-icon-and-count-select-multiple.html
http://localhost:4000/components/tree/test-toggle-icon-and-count-edit-auto.html
http://localhost:4000/components/tree/test-toggle-icon-and-count-edit-manually.html
- Should see plus-minus icon to parent nodes
- Should only do expand/collapse by icon (not by clicking on node text)
- Should show children count beside the node name text
- On `edit-auto/edit-manually` pages using `add/remove/update` node methods, so make sure show the right amount for children count

**Included in this Pull Request**:
- [x] A note to the change log.

### CODE REVIEW

File: src/components/tree/_tree.scss [hunk 1]
```diff
@@ -49,6 +59,20 @@
       margin: 0 5px 0 0;
       top: 2px;
       width: 15px;
+
+      &.rotate180-forward,
+      &.rotate180-backward {
+        animation-duration: 0.7s;
```
Comment (EdwardCoyle) [inline src/components/tree/_tree.scss:65]:
```suggestion
        animation-duration: 0.3s;
```
makes it end at the same time as the expand/collapse, what do you think?

Comment (tmcconechy) [reply src/components/tree/_tree.scss:65]: Nice then maybe you dont need to worry about the problem i saw with the children animating

**Comment by deep7102** [comment]
@tmcconechy @EdwardCoyle @jmacaluso711 
Updated now, for all above points found.

**Comment by jmacaluso711** [comment]
@deep7102 I'm seeing a bug that may or may not be related to the original issue. If I select a title and then use the toggle icons to open another tree, the selected title gets removed from the DOM. There doesn't seem to be any totally consistent pattern to recreate, but seems like it's mostly when a title is selected and you open/close one or more of the other trees:

IDS Enterprise (3)

**Comment by deep7102** [comment]
@jmacaluso711 good found, thanks!

**Comment by deep7102** [comment]
@tmcconechy @EdwardCoyle @jmacaluso711
Updated now, and ready to review.
